src/crypto/CHIPCryptoPALmbedTLS.cpp

```diff
@@ -1607,61 +1607,91 @@ CHIP_ERROR ExtractDNAttributeFromX509Cert(const uint8_t * oidAttribute, size_t o
     for (dnIterator = &mbed_cert.CHIP_CRYPTO_PAL_PRIVATE(subject); dnIterator != nullptr;
          dnIterator = dnIterator->CHIP_CRYPTO_PAL_PRIVATE(next))
     {
-        if (dnIterator != nullptr && dnIterator->CHIP_CRYPTO_PAL_PRIVATE(oid).CHIP_CRYPTO_PAL_PRIVATE(p) != nullptr &&
-            dnIterator->CHIP_CRYPTO_PAL_PRIVATE(oid).CHIP_CRYPTO_PAL_PRIVATE(len) == oidAttributeLen &&
-            memcmp(oidAttribute, dnIterator->CHIP_CRYPTO_PAL_PRIVATE(oid).CHIP_CRYPTO_PAL_PRIVATE(p),
-                   dnIterator->CHIP_CRYPTO_PAL_PRIVATE(oid).CHIP_CRYPTO_PAL_PRIVATE(len)) == 0 &&
-            dnIterator->CHIP_CRYPTO_PAL_PRIVATE(val).CHIP_CRYPTO_PAL_PRIVATE(p) != nullptr &&
-            dnIterator->CHIP_CRYPTO_PAL_PRIVATE(val).CHIP_CRYPTO_PAL_PRIVATE(len) == dnAttributeSize)
+        size_t oid_len  = dnIterator->CHIP_CRYPTO_PAL_PRIVATE(oid).CHIP_CRYPTO_PAL_PRIVATE(len);
+        uint8_t * oid_p = dnIterator->CHIP_CRYPTO_PAL_PRIVATE(oid).CHIP_CRYPTO_PAL_PRIVATE(p);
+        size_t val_len  = dnIterator->CHIP_CRYPTO_PAL_PRIVATE(val).CHIP_CRYPTO_PAL_PRIVATE(len);
+        uint8_t * val_p = dnIterator->CHIP_CRYPTO_PAL_PRIVATE(val).CHIP_CRYPTO_PAL_PRIVATE(p);
+
+        if (oid_p != nullptr && val_p != nullptr)
         {
-            // vendor id is of size 4, we should ensure the string is null terminated before passing in to strtoul to avoid
-            // undefined behavior
-            memcpy(dnAttribute, dnIterator->CHIP_CRYPTO_PAL_PRIVATE(val).CHIP_CRYPTO_PAL_PRIVATE(p), dnAttributeSize);
-            dnAttribute[dnAttributeSize] = 0;
-            VerifyOrExit(ArgParser::ParseInt(dnAttribute, id, 16), error = CHIP_ERROR_INTERNAL);
-            break;
+            bool isCommonNameAttr =
+                (oid_len == sizeof(sOID_AttributeType_CommonName)) && (memcmp(sOID_AttributeType_CommonName, oid_p, oid_len) == 0);
+            bool isMatterVidAttr = (oid_len == sizeof(sOID_AttributeType_MatterVendorId)) &&
+                (memcmp(sOID_AttributeType_MatterVendorId, oid_p, oid_len) == 0) && (val_len == kVIDandPIDHexLength);
+            bool isMatterPidAttr = (oid_len == sizeof(sOID_AttributeType_MatterProductId)) &&
+                (memcmp(sOID_AttributeType_MatterProductId, oid_p, oid_len) == 0) && (val_len == kVIDandPIDHexLength);
+
+            if (isMatterVidAttr || isMatterPidAttr)
```

Comment:
Is there any way to share this chunk of (nearly?) identical code with the OpenSSL impl?

Reply:
I agree this common logic could probably be factored-out between OpenSSL and mbedTLS, by doing a function in CHIPCryptoPAL that takes in Optional<> for VID/PID OID and for CN, and does the common logic to extract the VID PID.

Then this function would do the lib-specific cert parsing and delegate to that utility.

I think this could be done as a follow-up, though.

Reply:
Yes followup for that would be just fine.

Reply:
Added separate function `ExtractVIDPIDFromAttributeString()` in the `CHIPCryptoPAL.cpp`.